### ISSUE TITLE
nhl_xg: published boosters over-predict 25-30% for seasons through 2023-24

### ISSUE THREAD

**Issue by saiemgilani** [sportsdataverse/sportsdataverse-py#444]
The boosters `sportsdataverse.nhl.nhl_xg` downloads from the `nhl_xg_models` release (the 2026-04 fastRhockey artifacts) over-predict unblocked-shot goal probability by **25-30% for every season 2009-10 through 2023-24**. Ranking is fine; **levels are not**. Anything that SUMS `xg` is inflated for those seasons.

Diagnosed and evidenced in **sportsdataverse/fastRhockey-nhl-data#11** (merged 2026-09-02).

## Measurement

Committed 2026-04 boosters scored on `fastRhockey-nhl-data`'s `nhl/pbp/parquet` corpus with the frame the trainer builds today (a calibrated model gives goals/Î£xG = 1.0):

| variant | n shots | goals/Î£xG | rank AUC |
|---|---:|---:|---:|
| 5v5 | 1,724,290 | **0.7711** | 0.7783 |
| special teams | 349,232 | **0.7679** | 0.7601 |

Per season (5v5): 0.72-0.79 for every year through 2023-24, then **2024-25 0.949** and **2025-26 0.913** â€” much closer, though still low.

## Cause

The boosters were fit on a play-by-play corpus that carried no `MISSED_SHOT` events for those 15 seasons. Restricting today's frame to `SHOT|GOAL` moves goals/Î£xG to **1.0618** (5v5) / **1.0777** (ST) overall and to **0.99-1.02 for each individual season 2009-10 â€¦ 2023-24**, while the two seasons already near-calibrated on the full Fenwick frame blow out to 1.365 / 1.327. That is what identifies the training corpus.

This is **not** a feature-frame or Râ†”Python port defect â€” the two recipes were run on byte-identical input rows and returned the same rows and the same values on 32 of 33 shared features (the one divergence, `cross_ice_event`, is an R-side bug affecting 0.7% of rows). sdv-py's own `prepare_xg_features` agrees with both to 4 dp.

## What is affected

- `nhl_xg()` â€” the `xg` column it appends.
- Everything derived from it in this package: `nhl_gsax`, and any per-player/per-team xG or GAX aggregate.
- `load_xg_models()` / `ensure_xg_models()` â€” the other door to the same artifacts.
- **Known remaining exposure (not addressed by this issue):** the `xg` column carried in the published `nhl_pbp_full` release was produced by these same boosters, so a consumer reading that parquet directly â€” never importing sdv-py â€” gets the same inflated values **with no warning surface at all**. The honest fix there is a rebuild-and-republish of the dataset, which is a deliberate maintainer decision, not something a docstring can cover. Recording it here rather than papering over it.

## What this issue tracks

A `Warning`-style note in `nhl_xg()`'s long description stating the measured ratios and the date, so a consumer of the published asset can see the defect where they actually look (`help()` and the generated reference page). See #445.

## Removal trigger â€” delete the docstring note when ALL of:

1. A retrained xG booster passes its gates in `fastRhockey-nhl-data` and is committed (`models/xg_model_{5v5,st}.json`). Note the frozen floors (5v5 cv AUC â‰¥ 0.82, ST â‰¥ 0.81) are themselves anchored to the defective fit and are **not reachable** on the corrected corpus â€” they must be re-derived from a passing fit, not lowered; see fastRhockey-nhl-data#11 Â§3.
2. The **corpus bugs are fixed first**, or the re-derived floor is untrustworthy: `nhl/pbp/parquet/` is missing season 2014-15, and `nhl/pbp/full/{rds,parquet}` is missing 2013-14 and duplicates 2018-19 (its `play_by_play_2014â€¦2018` files are named by season START year while every other file uses END year).
3. The `sportsdataverse-data` release **`nhl_xg_models`** is re-uploaded with the new boosters â€” until that happens sdv-py consumers still download the old ones, so the note still applies even after the retrain lands.
4. Re-measured: `nhl_data_build.xg_parity.artifact_calibration` (added in fastRhockey-nhl-data#11) reports goals/Î£xG within a band of 1.0 on the current corpus.

Anyone can re-check the state at any time by summing `xg` for a season and comparing against actual goals â€” a corrected booster gives â‰ˆ 1.0.

**Comment by saiemgilani** [comment]
**The stated root cause does not survive checking, and I think the removal trigger on this issue is aimed at the wrong thing.**

This issue records the cause as: *"the boosters' training corpus carried no `MISSED_SHOT` events for the affected seasons."* Three measurements, none of which support that.

### 1. The corpus has missed shots in every season

`nhl/pbp/parquet/*.parquet`, the exact glob the trainer reads (2010–2026, 16 files):

```
season     rows      SHOT   MISSED    GOAL   missed% of Fenwick
2010    400,512    73,090   30,999   7,716        27.7%
2014    853,324    73,120   31,431   7,506        28.0%
2020    711,747    68,606   28,147   7,297        27.1%
2023  1,103,989    79,059   34,625   9,027        28.2%
2026  1,100,073    69,507   41,811   8,836        34.8%
```

No season is missing them; the share rises 27% → 35% but is never absent. The trainer also selects on `_FENWICK = ["SHOT", "MISSED_SHOT", "GOAL"]`, so the code was never excluding them either.

### 2. The published boosters were themselves fit on Fenwick

An XGBoost JSON records its training label mean in `base_score`. Downloaded from the `nhl_xg_models` release (assets dated 2026-06-27):

| | base_score | interpretation |
|---|---|---|
| published `xg_model_5v5.json` | **0.07368** | ≈ the Fenwick goal rate |
| published `xg_model_st.json` | **0.10533** | ≈ the ST Fenwick rate |
| corpus: goals / Fenwick | 0.0695 | what a correct fit sits near |
| corpus: goals / shots-on-goal | 0.0977 | what a MISSED-SHOT-excluding fit would sit near |

If the boosters had been trained without missed shots, `base_score` would sit near **0.0977**. It sits near the Fenwick rate. Both published artifacts carry all 36/38 features including `era_2025_on`, and are identical in these parameters to the locally committed copies.

### 3. A base-rate mechanism was the obvious alternative, and it is also wrong

If a model learned `P(goal | shot on goal)` and were applied to all Fenwick, the over-prediction would be exactly the base-rate ratio: pooled **0.712** predicted, against **0.771 / 0.768** observed — the right magnitude, so it looked promising. But per season it predicts the two most recent seasons should be the **worst** (2025 → 0.659, 2026 → 0.652), and this issue records them as the **best** (0.949, 0.913). The direction contradicts, so I am discarding it rather than reporting the pooled agreement.

### What this means for the trigger

The four-part removal trigger here starts with "corpus bugs fixed first." **There is no corpus bug of the kind described**, so that step cannot be satisfied and the retrain behind it is aimed at a cause that is not present. That is consistent with what the retrain actually did: `cv_auc` 0.7786 (5v5) and 0.7569 (ST) against the frozen 0.82 gate — a model fit on a corpus that was never the problem, correctly failing to beat the incumbent.

**The 25–30% over-prediction is real and reproduced; its cause is now unidentified.** I would not retrain again until it is, and I would not treat the AUC gate as the obstacle.

Next check I would run (not done yet): score the published booster on a held-out season and compare the predicted distribution against observed goals *by shot-distance bucket*. If the over-prediction is uniform it points at levelling/calibration; if it concentrates in one distance band it points at a feature, most likely the coordinate handling that PWHL's xG had (58.7% of shots at distance 0).

Leaving the issue open; only correcting the diagnosis so the next person does not spend the effort on the corpus.

**Comment by saiemgilani** [comment]
Follow-up: built leave-one-season-out CV for the 5v5 booster and ran a feature/dtype search against it. **The shipped 36-feature recipe is at the ceiling of what this data supports** — seven directions, none beat it.

## LOSO results (16 seasons; early stopping on a train slice, held-out season never touched)

| direction | LOSO mean AUC | vs shipped |
|---|---:|---:|
| **shipped 36 features, depth 4** | **0.7769** | — |
| depth 8 | 0.7772 | +0.0003 |
| depth 6 | 0.7771 | +0.0002 |
| + raw x/y, abs_y, score diff, period, game seconds, home flag | 0.7766 | −0.0003 |
| + shooter/goalie talent as out-of-fold shrunk rates | 0.7767 | −0.0002 |
| + goalie id (native categorical) | 0.7722 | −0.0047 |
| **21 one-hots → 2 native categoricals** | 0.7685 | −0.0084 |
| + shooter id (native categorical) | 0.7671 | −0.0098 |

Doubling tree depth buys +0.0007, so the model is not capacity-limited. Feature importance agrees: `empty_net` carries **38.3% of gain** (a near-deterministic goal flag, not shot quality), `shot_distance` 5.3%, `shot_angle` not in the top 12, and five features are never split on (`between_legs`, `cradle`, `last_stop`, `last_penalty`, `last_goal`).

Entity identity fails the way LOSO is designed to expose: raw high-cardinality ids memorise levels that do not survive a season boundary. Encoded properly as out-of-fold shrunk rates they carry real signal (**0.6059 alone**) but add nothing on top of the 36 — the geometry already captures it.

**Conclusion: ~0.78 is the honest ceiling for 5v5 unblocked shots on NHL public play-by-play**, consistent with public models. The missing signal is pre-shot passing sequence, screens and traffic, which the feed does not contain. Reaching 0.80 needs different data, not a better fit — and the 0.82 `cv_auc` gate on `nhl_model_01_xg_5v5` cannot be met by any honest fit on this corpus.

## Two leaks found on the way, both worth knowing

1. **`home_score` / `away_score` are POST-event.** On 2024, **8,590 of 8,772 goal rows (97.9%)** already include the goal being scored. A naive `score_diff` built from them scored **LOSO 0.8495** against the 0.7769 control — it reads as clearing the target and is a label tell. Use the score shifted within the game. There is also an `xg` column sitting in the pbp frame (the model's own output), a live hazard for anyone building features off it.
2. **Leave-one-out target encoding INVERTS the leak rather than removing it.** `(s − y_i + k·p0)/(n − 1 + k)` subtracts the row's own outcome, so within a shooter every goal row gets a systematically lower rate; the model learns "lower rate → goal" and is exactly backwards on test. Measured: base36 **0.7808 → 0.5882**, rates alone **0.5178**, with both fits running to the round cap instead of early-stopping. Out-of-fold encoding fixes it (rates alone 0.5178 → **0.6059**, early stop at 94 rounds).

## Bearing on this issue

The calibration measurements matter more than the AUC search. **Every held-out season lands at goals/Σxg between 0.95 and 1.05, mean ≈ 1.00**, against the published booster's 0.771. Combined with the earlier finding that the published `base_score` (0.0737) matches the Fenwick rate rather than the shots-on-goal rate, the 25–30% over-prediction is a property of the published artifact — not the corpus, the recipe, or the training code.

So the fix remains **retrain and republish**, with no corpus repair, and the AUC gate should be re-derived from an honest LOSO fit rather than from the 0.8322 the miscalibrated artifact reported.
